src/OaiPmh/Harvester.php

```diff
@@ -220,6 +218,8 @@ protected function sendRequest($verb, $params = [])
 
     /**
      * Load date granularity from the server.
+     * 
+     * @deprecated Moved to VuFindHarvest\OaiPmh\Harvester::storeDateSettings()
```

Comment:
I think you can safely remove this method rather than just deprecating it. We'll need to release this as a major version due to backward-breaking changes, and I don't think there's any value in retaining the legacy function.

Reply:
Great. This was my instinct as well, especially as I don't see any way for folks to dynamically extend this Harvester class in their own implementations.

Thinking further, while I've preserved (moved) the logic to set the granularity property, it does not look like the property itself is actually used anymore. It used to be part of the final date calculation for last_harvest.txt (in [getRecords()](https://github.com/vufind-org/vufindharvest/blob/dev/src/OaiPmh/Harvester.php#L271)), but that's getting replaced as a fundamental part of this PR. I think the whole property could be dropped. Can you see any forward-compatible justification to keep it?

Reply:
In theory, it is possible to dynamically extend the Harvester in VuFind (you can override the configuration for the harvest command and replace it with your own subclass and factory), but in practice it would be a lot of work since you'd have to override a lot of pieces to make it work. This perhaps highlights the fact that our factory could be made more flexible by replacing some hard-coded class names with configurable properties... but that's another conversation for another thread. :-)

We might need to think a little more about the granularity setting. That is currently a setting that gets loaded from oai.ini or console parameters (see https://github.com/vufind-org/vufindharvest/blob/dev/src/OaiPmh/HarvesterCommand.php#L226-L230). The original purpose of this was that if you knew and specified the granularity in advance, we could bypass the Identify call. Now that we have a more frequent reason to call Identify, the value of that is significantly reduced, but I think there are potentially still situations where it applies -- i.e. if the granularity is specified in advance and we're given an "Until" date, I think it's still possible to proceed without calling Identify.

Reply:
Yes, this is interesting. I'm still trying to untangle exactly what we need to be doing with that granularity value (no matter how it's sourced). Previously we were using it to convert a non-UTC timestamp from record headers into a UTCdatetime format that can be used in last_harvest.txt. Now however, the last_harvest.txt value is based on a server datestamp that's already UTC, or (seemingly very rarely) a direct user-supplied "until" value. I'm wondering if extra care and validation is actually needed to convert all selective harvesting values into the right granularity. The OAI spec just seems to indicate that "from" and "until" values need to be UTCdatetime format, but I don't see any requirement about exactly matching the server's granularity setting for those. I though this granularity was just an indication of what behavior to expect, as opposed to a strict input format requirement.

As it stands right now in the PR the "until" value is based on last_harvest.txt, which is now tied to the previous harvest's server identify responseDate value. This is always UTC in _second_ granularity. I wonder if later passing this as a "from" value to an OAI host with "day" granularity will cause issues. If so, we probably need some sort of date validate/convert method to check the format against the granularity. I suppose that any user-supplied "from" and "until" values could also be passed though this.

Reply:
Ahh, you're right, it does appear that maybe the whole granularity thing can simply be removed at this point. It looks to me like the purpose of it was to ensure that dates we generated were consistent with dates received from the server... but since all that logic is gone, it no longer matters.

Reply:
Thanks for checking my reasoning here! While this may contradict my “let the spec be the main guide” tone from earlier comments, I’m still mindful of that one case where an OAI-PMH host with day granularity starts getting a “from” value with second granularity. While the spec indicates that this certainly should not matter, one takeaway from these conversations is that OAI-PMH hosts may behave inconsistently. Reflecting more -- if there is an edge case where this actually causes a problem, we could be introducing a regression.

While enforcing UCT the previous implementation did also flip a system-provided date into YYYY-MM-DD format before writing the value to last_harvest.txt when day granularity was detected. We could (should?) continue to do that with little risk. I had considered some sort of date format checker/filter for all values sent as “from” and “until” params, but given the way direct user inputs could be used for those (with varying formats and timezone encodings), that seems like a risky proposition. The new Identify responseDate format is much more predictable though, and so a granularity check/conversion on it seems doable within a context likes launch(), where we know `$explicitHarvestEndDate` is only being used to source last_harvest.txt data:

```
if (empty($this->harvestEndDate)) {
    $explicitHarvestEndDate = $this->getIdentifyResponse()->responseDate;
    // Backwards compatible support for any OAI-PMH hosts that may require
    // day granularity format for all date params.
    $granularity = $this->granularity == 'auto' ? $this->getIdentifyResponse()->granularity : $this->granularity;
    if ($granularity == 'YYYY-MM-DD') {
      $explicitHarvestEndDate = (new \DateTime($explicitHarvestEndDate))->format('Y-m-d');
    }
} else {
    $explicitHarvestEndDate = $this->harvestEndDate;
}
```
I don't want to over-analyze an edge case here, but also want to be respectful of these nuances from the previous implementation if needed.

Reply:
Good point. My original thought was that if we're feeding back data the system provides to us, it should be able to understand its own output as input, but you're right that we can't really make this assumption -- the system could potentially be able to report its own clock time in full granularity while serving up data with lesser granularity... and the spec says that an OAI server MUST throw an error if it receives from/until parameters in the wrong granularity. Thus, I believe you are correct that we still need the granularity check. Your code above makes sense to me, though given the predictability of the formats, maybe it would be better to trim the UTCdate with a substr() rather than doing a full date conversion. That would reduce chances of something unpredictable happening and also require less processing. (Obviously it could cause trouble if the server returns malformed date values, but I think that's going to cause trouble in any situation).

Reply:
Great! I'm glad we talked though that detail. https://github.com/vufind-org/vufindharvest/pull/8/commits/314bf0a1c86a769004f3217c99b27503efcde005 adds the granularity check. I think this means that we also have to maintain some test coverage for differing granularity, but that should not be an issue.